CONTRIBUTING.md

```diff
@@ -1,143 +1,269 @@
-1. Type `./paper patch` in git bash to apply the changes from upstream
-1. cd into `Paper-Server` for server changes, and `Paper-API` for API changes
+To get started with PRing changes, you'll need the following software, most of
+which are all available in (most) package managers such as `apt` (Debian /
+Ubuntu), `homebrew` (macOS / Linux), and more:
 
-These directories aren't git repositories in the traditional sense:
+- `git` (package `git` everywhere);
+- `patch` (often package `patch`);
+- A Java 8 or later JDK (packages vary, use Google/DuckDuckGo/etc.).  
+If you need one, you can find them on [AdoptOpenJDK](https://adoptopenjdk.net/).
+- `maven` (often package `maven`; can be found on
+[Apache's site](https://maven.apache.org/download.cgi) too).
 
-- Every single commit in Paper-Server/API is a patch. 
-- 'origin/master' points to a directory similar to Paper-Server/API but for Paper
-- Typing `git status` should show that we are 10 or 11 commits ahead of master, meaning we have 10 or 11 patches that Paper and Spigot don't
-  - If it says something like `212 commits ahead, 207 commits behind`, then type `git fetch` to update spigot/paper
+If you're on Windows, check
+[the section on WSL](#patching-and-building-is-really-slow-what-can-i-do).
 
-## Adding Patches
-Adding patches to Paper is very simple:
+If you're compiling with Docker, you can use the
+[`adoptopenjdk`](https://hub.docker.com/_/adoptopenjdk/) images like so:
 
-1. Modify `Paper-Server` and/or `Paper-API` with the appropriate changes
-1. Type `git add .` to add your changes
-1. Run `git commit` with the desired patch message
-1. Run `./paper rebuild` in the main directory to convert your commit into a new patch
-1. PR your patches back to this repository
+```console
+# docker run -it -v "$(pwd)":/data --rm adoptopenjdk:8-jdk-hotspot bash
+Pulling image...
 
-Your commit will be converted into a patch that you can then PR into Paper
+root@abcdefg1234:/# javac -version
+javac 1.8.0_252
+```
 
-## Help! I can't find the file I'm looking for!
-By default, Paper (and upstream) only import files that we make changes to.
-If you would like to make changes to a file that isn't present in Paper-Server's source directory, you
-just need to add it to our import script to be ran during the patch process.
+## Understanding Patches
 
-1. Save (rebuild) any patches you are in the middle of working on!
-1. Identify the names of the files you want to import.
-   - A complete list of all possible file names can be found at ```./work/Minecraft/$MCVER/spigot/net/minecraft/server```
-1. Open the file at `./scripts/importmcdev.sh` and add the name of your file to the script.
-1. Re-patch the server `./paper patch`
-1. Edit away!
+Paper is essentially a set of patches on top of Spigot. These patches are split
+into different directories which target a certain part of the code. These
+directories are:
+
+- `Paper-API` - Modifications to `Spigot-API`/`Bukkit`;
+- `Paper-MojangAPI` - An API for
```

Comment:
"Paper is essentially a set of patches on top of Spigot" - MojangAPI doesn't really fit under that description